- [Description](#Description)
- [Installation](#Installation)
- [Usage](#Usage)

## Description

## Installation

Use the [Conda](https://docs.conda.io/en/latest/) package manager and [BioConda](https://bioconda.github.io/index.html) channel to install TBvar.

If you do not have conda installed do the following:

```bash
# Download Conda installer
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
# Set permissions
chmod -X Miniconda3-latest-Linux-x86_64.sh
# Install
bash Miniconda3-latest-Linux-x86_64.sh
```

Set up channels:

```bash
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
```

Get **TBvar** pipeline:

```bash
git clone https://github.com/dbespiatykh/TBvar.git
```

Get [SpolLineages](https://github.com/dcouvin/SpolLineages):

```bash
cd TBvar && mkdir bin && cd $_
git clone https://github.com/dcouvin/SpolLineages.git
cd SpolLineages/Binary_Mask2
gcc Mask2.c -o Mask2
```
Install required dependencies:

```bash
cd TBvar
conda env create --file environment.yml
```

## Usage

Activate **TBvar** environment:

```bash
conda activate TBvar
cd TBvar
```

:point_right: In project folder make `FASTQ` folder:

```bash
mkdir FASTQ
```

and move to it (symlink) your paired- or single-end [FASTQ](https://en.wikipedia.org/wiki/FASTQ_format) reads. Paired-end FASTQ suffix should be: `_1.fastq.gz` and `_2.fastq.gz`, for single-end it should be: `_1.fastq.gz`.

> If you are using both PE and SE reads, firts time you run the pipleine it will throw an error `AmbiguousRuleException:...`, just ignore it and run it again.

Project folder should have the following structure:

```bash
ðŸ“‚TBvar/
â”œâ”€â”€ FASTQ
â”‚   â”œâ”€â”€ SampleA_1.fastq.gz
|   â”œâ”€â”€ SampleA_2.fastq.gz
|   â”œâ”€â”€ SampleB_1.fastq.gz
|   â””â”€â”€ ...
â”œâ”€â”€ LICENSE
â”œâ”€â”€ README.md
â”œâ”€â”€ Snakefile
â”œâ”€â”€ bin
â”‚   â””â”€â”€ SpolLineages
â”‚       â”œâ”€â”€ Binary_Mask2
|       â”œâ”€â”€ ...
|       ...
â”œâ”€â”€ environment.yml
â”œâ”€â”€ envs
â”‚   â”œâ”€â”€ barcoding.yaml
â”‚   â”œâ”€â”€ bcftools.yaml
â”‚   â”œâ”€â”€ rtg_tools.yaml
â”‚   â”œâ”€â”€ snpit.yaml
â”‚   â”œâ”€â”€ spollineages.yaml
â”‚   â”œâ”€â”€ spotyping.yaml
â”‚   â””â”€â”€ tb_profiler.yaml
â”œâ”€â”€ ref
â”‚   â”œâ”€â”€ NC_000962.3.dict
â”‚   â”œâ”€â”€ NC_000962.3.fna
â”‚   â”œâ”€â”€ NC_000962.3.fna.amb
â”‚   â”œâ”€â”€ NC_000962.3.fna.ann
â”‚   â”œâ”€â”€ NC_000962.3.fna.bwt
â”‚   â”œâ”€â”€ NC_000962.3.fna.fai
â”‚   â”œâ”€â”€ NC_000962.3.fna.pac
â”‚   â”œâ”€â”€ NC_000962.3.fna.sa
â”‚   â””â”€â”€ tbdb
â”‚       â”œâ”€â”€ tbdb.ann.txt
â”‚       â”œâ”€â”€ tbdb.barcode.bed
â”‚       â”œâ”€â”€ tbdb.bed
â”‚       â”œâ”€â”€ tbdb.confidence.csv
â”‚       â”œâ”€â”€ tbdb.csv
â”‚       â”œâ”€â”€ tbdb.dr.json
â”‚       â”œâ”€â”€ tbdb.fasta
â”‚       â”œâ”€â”€ tbdb.fasta.fai
â”‚       â”œâ”€â”€ tbdb.gff
â”‚       â”œâ”€â”€ tbdb.version.json
â”‚       â””â”€â”€ tbdb.watchlist.csv
â””â”€â”€scripts
    â”œâ”€â”€ aggregate.py
    â”œâ”€â”€ concat.py
    â”œâ”€â”€ gather_spol.py
    â”œâ”€â”€ read_vcf_v3.py
    â””â”€â”€ rtg_concat.py
```

Run pipeline:

```bash
snakemake --conda-frontend mamba --use-conda -j 48 -c 48 --max-threads 48 -k --rerun-incomplete
```

It is recommended to use dry run if you are running pipeline for the first time, to see if everything is in working order, for this you can use `-n` flag:

```bash
snakemake --conda-frontend mamba --use-conda -j 48 -c 48 --max-threads 48 -k --rerun-incomplete -np
```
